design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,249 @@ status: draft
+capabilities - navigating an object's capabilities and managing the object's lifetime. All COM
+interfaces derive from IUnknown and all objects in any COM based system must implement these methods.
+
+``` cpp
+// IUnknown GUID identifier: 00000000-0000-0000-C000-000000000046
+struct IUnknown
+{
+    int32_t  QueryInterface(guid const& id, void** object);
+    uint32_t AddRef();
+    uint32_t Release();
+};
+```
+
+### QueryInterface
+
+Given any COM interface pointer, you can query for any other COM interface via IUnknown's
+QueryInterface method. If the referenced COM object supports the interface being queried, a pointer
+to that COM interface is returned to the caller. Otherwise, an error code indicating that the
+requested interface is not supported.
+
+As stated above, interfaces are identified by a GUID. This GUID is passed by reference as the first
+parameter of QueryInterface. The pointer to the requested interface - if available - is returned via
+a void** out parameter. It is the responsibility of the calling code to cast this pointer to the
+correct COM interface pointer type.
+
+Methods on COM interfaces (including xlang ABI interfaces) never throw C++ exceptions, due to the
+need for a stable contract across compilers. QueryInterface returns a 32-bit signed integer as an
+error code. On Windows, this code is known as an
+[HRESULT](https://docs.microsoft.com/en-us/windows/desktop/com/error-handling-in-com).
+Negative HRESULT values indicate errors while non-negative HRESULT values (including zero) indicate
+success. QueryInterface in particular should always return 0 on success.
+
+If the requested interface is not supported by the object, QueryInterface returns the error code
+value 0x80004002 (E_NOINTERFACE).
+
+### AddRef and Release
+
+COM objects manage their own lifetime via an internal reference count. When this count goes to zero,
+the COM object cleans itself up. IUnknown provides methods to increment and decrement the object's
+reference count. If a language projection makes a copy of a COM interface pointer, it must call
+AddRef. When client code is finished using a given pointer to a COM interface, the language
+projection must call Release.
+
+The return value of these function is specified to be the new reference count of the object. However
+the COM standard itself warns that this value is intended to be used only for test purposes. In
+practice on Windows, the value returned from these two functions is not guaranteed to be the object's
+true reference count. For example, COM object can choose to implement an "tear-off" interfaces.
+A tear-off interface is implemented on a separate object, which is constructed when the interface
+in question is queried. When AddRef or Release is called on the tear-off interface, it will return
+the reference count of the tear off interface object *not* the reference count of the original
+object.
+
+## IXlangObject
```

Comment:
Should IXlangObject not be named with Object since it can be used to represent interfaces as well as objects?

Reply:
Yeah understood but the terms across languages are of course overloaded and confusing which is why I'm wondering if there is something better (there might not be). I *think* we are probably fine since Java uses Object as its root and ObjC has NSObject so there is some prior art.

Reply:
It's worth keeping in mind that this is infrastructure only seen by projection authors. End users won't deal with the interface, for the most part, so the naming here does not need to be a coherent ideal name for all possible languages.